File: tests/e2e/vec_add.c
#include <stdio.h>

extern void vec_add(long N, float *out, float *a, float *b);

int main() {
    float a[5] = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f};
    float b[5] = {10.0f, 20.0f, 30.0f, 40.0f, 50.0f};
    float out[5] = {0};
    vec_add(5, out, a, b);
    printf("%.0f %.0f %.0f %.0f %.0f\n", out[0], out[1], out[2], out[3], out[4]);
    return 0;
}

File: tests/e2e/vec_add.py
# RUN: uv run xdsl-exo -o %t.mlir %s
# RUN: mlir-opt --convert-arith-to-llvm --convert-cf-to-llvm --convert-func-to-llvm --reconcile-unrealized-casts %t.mlir | mlir-translate --mlir-to-llvmir | llc -filetype=obj -o %t.o
# RUN: clang %t.o %S/vec_add.c -o %t.bin -lm
# RUN: %t.bin | filecheck %s

from __future__ import annotations

from exo import *


# CHECK: 11 22 33 44 55
@proc
def vec_add(N: size, out: f32[N] @ DRAM, a: f32[N] @ DRAM, b: f32[N] @ DRAM):
    for i in seq(0, N):
        out[i] = a[i] + b[i]
